contracts/src/token/erc20/utils/safe_erc20.rs

```diff
@@ -0,0 +1,153 @@
+    #[derive(Debug)]
+    #[allow(missing_docs)]
+    error SafeErc20FailedDecreaseAllowance(address spender, uint256 currentAllowance, uint256 requestedDecrease);
+}
+
+/// A SafeErc20 error
+#[derive(SolidityError, Debug)]
+pub enum Error {
+    /// Error type from [`Erc20`] contract [`erc20::Error`].
+    Erc20(erc20::Error),
+    /// An operation with an ERC-20 token failed.
+    SafeErc20FailedOperation(SafeErc20FailedOperation),
+    /// Indicates a failed `decreaseAllowance` request.
+    SafeErc20FailedDecreaseAllowance(SafeErc20FailedDecreaseAllowance),
+}
+
+/// Wrappers around ERC-20 operations that throw on failure (when the token
+/// contract returns false). Tokens that return no value (and instead revert or
+/// throw on failure) are also supported, non-reverting calls are assumed to be
+/// successful.
+/// To use this library you can add a `using SafeERC20 for IERC20;` statement to
+/// your contract, which allows you to call the safe operations as
+/// `token.safeTransfer(...)`, etc.
+pub trait SafeErc20 {
+    /// The error type associated to this Safe ERC-20 trait implementation.
+    type Error: Into<alloc::vec::Vec<u8>>;
+
+    /// Transfer `value` amount of `token` from the calling contract to `to`. If
+    /// `token` returns no value, non-reverting calls are assumed to be
+    /// successful.
+    fn safe_transfer(
+        &self,
+        to: Address,
+        value: U256,
+    ) -> Result<(), Self::Error>;
+}
+
+impl SafeErc20 for Erc20 {
+    type Error = Error;
+
+    fn safe_transfer(&self, to: Address, value: U256) -> Result<(), Error> {
+        type TransferType = (SOLAddress, Uint<256>);
+        let tx_data = (to, value);
+        let data = TransferType::abi_encode_params(&tx_data);
+        let hashed_function_selector =
+            function_selector!("transfer", Address, U256);
+        // Combine function selector and input data (use abi_packed way)
+        let calldata = [&hashed_function_selector[..4], &data].concat();
+
+        self.call_optional_return(calldata)
+    }
+}
+
+/// NOTE: Implementation of [`TopLevelStorage`] to be able use `&mut self` when
+/// calling other contracts and not `&mut (impl TopLevelStorage +
+/// BorrowMut<Self>)`. Should be fixed in the future by the Stylus team.
+unsafe impl TopLevelStorage for Erc20 {}
+
+impl Erc20 {
+    /// Imitates a Solidity high-level call (i.e. a regular function call to a
+    /// contract), relaxing the requirement on the return value: the return
+    /// value is optional (but if data is returned, it must not be false).
+    /// @param token The token targeted by the call.
+    /// @param data The call data (encoded using abi.encode or one of its
+    /// variants).
+    ///
+    /// This is a variant of {_callOptionalReturnBool} that reverts if call
+    /// fails to meet the requirements.
+    fn call_optional_return(&self, data: Vec<u8>) -> Result<(), Error> {
+        match RawCall::new()
```

Comment:
Certain things are unclear to me here as the docs aren't clear enough:
1. How can one get the address of a contract (in this case the ERC20 token) without passing it explicitly? Similar to how in Solidity one does `address(contract)`.
2. My intuition says that [RawCall + limit_return_data](https://docs.arbitrum.io/stylus/reference/rust-sdk-guide#rawcall-and-unsafe-calls) is more performant here, but I'm a bit confused by this sentence in the docs: ``Note that the [call](https://docs.rs/stylus-sdk/latest/stylus_sdk/call/struct.RawCall.html#method.call) method is `unsafe` when reentrancy is enabled.``. May be a dumb question, but better safe than sorry - does this mean the `call` is safe to use here, but we let the caller of the function worry about reentrancy?

Reply:
Replied in the other comment about different approach